### ISSUE TITLE
AOT FLOPS Calculation Doesn't Work, HLO FLOPS Calculation Possibly Wrong

### ISSUE THREAD

**Issue by foainla** [google/jax#17912]
### Description

The ahead of time lowering and compilation method for calculating the FLOPS described in (https://jax.readthedocs.io/en/latest/aot.html) returns -1 when applied to a flax.linen.Module involving convolution. The other way of calculating FLOPS using hlo_module_cost_analysis works but gives a different answer than expected.

I applied a Conv block with 2 features with (2,2) kernel size, no bias, "same" padding, (1,1) strides to a (1,4,4,1) array both in flax and keras, then calculated flops used for each. Keras flop calculation would return 256 which makes sense (2 kernels * 16 output cells * (4 mul + 4 sum) = 256). However, both the JAX internal flop calculation tool and the flax flop calculation from the tabulate function return 196 as flops used. How can that be? If we were to dissect the number 196 by dividing it by 2 kernels and 16 output cells, we would get 6.125 flops per cell. The only way I think this would be possible is 4 mul and 2 1/8th add per cell with some sort of vectorization optimization going on? This discrepancy between keras and flax flop count doesn't occur if the padding is "valid".

The google colab link includes this short display:
https://colab.research.google.com/drive/1ClNx5FWMIvf_H_jGVV4Z2T1_9J0ZmNlF?authuser=0#scrollTo=e8mzJ2rvvgRL

### What jax/jaxlib version are you using?

jax v0.4.17, jaxlib v0.4.14

### Which accelerator(s) are you using?

GPU T4

### Additional system info

Python Version: 3.10.12, OS: Linux Ubuntu 22.04, Flax v0.7.5

### NVIDIA GPU info

Wed Oct  4 06:06:20 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   51C    P0    28W /  70W |  12221MiB / 15360MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+

**Comment by romanngg** [comment]
The discrepancy appears to be the cost of convolving the padded entries (https://github.com/google/flax/issues/3392#issuecomment-1747553622)

**Comment by foainla** [comment]
Thank you, that part makes sense. However, why does the ahead of time lowering and compiling cost analysis return -1 as the number of flops?

**Comment by romanngg** [comment]
I don't know the reason (Hope JAX/XLA team could elaborate), but know that compiled cost analysis indeed often doesn't work or works incorrectly on CPU and GPU (see https://github.com/google/jax/issues/16008, Google-internal: b/202218145)
